Proposal: Dapr Error Handling Codes (#21)
dapr/proposals: 000X-B-C-I-R-S-error-handling-codes.md

// different programming environments, including REST APIs and RPC APIs.
message Status {
  // A simple error code that can be easily handled by the client. The
  // actual error code is defined by `google.rpc.Code`.
  int32 code = 1;

  // A developer-facing human-readable error message in English. It should
  // both explain the error and offer an actionable resolution to it.
  string message = 2;

  // Additional error information that the client code can use to handle
  // the error, such as retry info or a help link.
  repeated google.protobuf.Any details = 3;
}
```

Here is one of the possible details that can be added to the above
error structure. This is defined in the [error_details.proto Protobuf](https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto) 

```go
message ErrorInfo {
  // The reason of the error. This is a constant value that identifies the
  // proximate cause of the error. Error reasons are unique within a particular
  // domain of errors. This should be at most 63 characters and match a
  // regular expression of `[A-Z][A-Z0-9_]+[A-Z0-9]`, which represents
  // UPPER_SNAKE_CASE.

>   string reason = 1;

Reviewer: is there a particular max length that is recommended for this?

Author: 63 characters appears to be the max length. I haven't found out why, but to be on the safe side we should stick to it.

  // The logical grouping to which the "reason" belongs. The error domain
  // is typically the registered service name of the tool or product that
  // generates the error. Example: "pubsub.googleapis.com". If the error is
  // generated by some common infrastructure, the error domain must be a
  // globally unique value that identifies the infrastructure. For Google API
  // infrastructure, the error domain is "googleapis.com".

>   string domain = 2;

Reviewer: is it going to be only `dapr.io` here? And if so do we need the domain field?

Author: For now we're planning it to be `dapr.io`.

Reviewer 2: ErrorInfo is not out structure, it is Google's gRPC and we will pass `dapr.io`, so SDKs can know which ones they care about in the array of ErrorInfo.

  // instances that can be created in a single (batch) request.
  map<string, string> metadata = 3;
}
```

### Error Status
The properties of the **google.rpc.Status** will be populated as following:

- **Code**  - Protocol level error code. These could be either gRPC or HTTP error codes. See (gRPC Codes ProtoBuf)[https://github.com/googleapis/googleapis/blob/master/google/rpc/code.proto]
   
   Example: "InvalidArgument Code = 3", "Internal Code = 13"

- **Message**  - Error message.
- **Details**  - A set of standard error payloads for error details. These list can be found in [Error Details](https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto)
   
   Example: "ErrorInfo", "ResourceInfo"

Below is partial table of the Standard Error code provided by gRPC and how they map to HTTP error codes. The entire list can found in the following links:
- [Google API Error Handling]https://cloud.google.com/apis/design/errors#handling_errors
- (gRPC Codes ProtoBuf)[https://github.com/googleapis/googleapis/blob/master/google/rpc/code.proto]

|HTTP | gRPC      | Description |
|---- | ----------- | ----------- |

> |200  |	OK		  | No error. |

Reviewer: Probably list all the current set of HTTP return codes and their mappings that dapr currently returns.

Author: At the moment, we don't have a mapping to Dapr specific errors. As we discussed in the review call, we should know more as we dig deeper.

|404  |	NOT_FOUND |	A specified resource is not found. |
|409  |	ABORTED	| Concurrency conflict, such as read-modify-write conflict. |

### ErrorInfo
The properties of the **type.googleapis.com/google.rpc.ErrorInfo** will be populated as following:

- **Reason** - A combination of prefix from prefix of the table below plus the error condition code.
    
    Example: "DAPR_STATE_" + "ETAG_MISMATCH"

- **Domain** - With the value `dapr.io`.

- **Metadata** - A key/value map/dictionary data relevant to the error condition.

### ResourceInfo
The properties of the **type.googleapis.com/google.rpc.ResourceInfo** will be populated as following:

- **ResourceType** - The building block type.
   
   Example: "StateStore"

- **ResourceName** - The component name.
   
   Example: "REDIS"

- **Owner**   - The owner of the component.

- **Description** - Resource descrpition or error details.

### Error Details Prefixes
The following tables shows the propsosed error codes prefixes used in the **reason** for the **google.rpc.ErrorInfo** for various Dapr building blocks:

**INIT**
| Dapr Module | Prefix |
| ----------- | ----------- |

> | CLI         | DAPR_CLI_INIT_* |

Reviewer: Can we leave it at `DAPR_INIT_*`? We do have a separate dapr `CLI` and `daprd` is a binary and not a CLI.

Author: As we dig deeper into each building block we can redefine some of these.

> | Self-hosted | DAPR_SELF_HOSTED_INIT_* |

Reviewer: I think, these should be completely platform agnostic.

| K8S         | DAPR_K8S_INIT_* |
| Invoke      | DAPR_INVOKE_INIT_* |

**RUNTIME**
| Dapr Module | Prefix |
| ----------- | ----------- |

> | CLI         | DAPR_CLI_RUNTIME_* |

Reviewer: what exactly does CLI RUNTIME mean here? 
Probably just `DAPR_RUNTIME_*`? 
Or Probably just `DAPR_EXEC_*`?

Author: `CLI RUNTIME` would mean something like `INVOKE`, etc... As we dig deeper into each building block we can redefine some of these.

> | Self-hosted | DAPR_SELF_HOSTED_RUNTIME_* |

Reviewer: remove platform specific codes.

| dapr-2-dapr(HTTP) | DAPR_HTTP_RUNTIME_* |

> | dapr-2-dapr(gRPC) | DAPR_GRPC_RUNTIME_* |

Reviewer: the other way around 
```
DAPR_RUNTIME_HTTP_*
DAPR_RUNTIME_GRPC_*
```

  // UPPER_SNAKE_CASE.
  string reason = 1;

  // The logical grouping to which the "reason" belongs. The error domain
  // is typically the registered service name of the tool or product that
  // generates the error. Example: "pubsub.googleapis.com". If the error is
  // generated by some common infrastructure, the error domain must be a
  // globally unique value that identifies the infrastructure. For Google API
  // infrastructure, the error domain is "googleapis.com".
  string domain = 2;

  // Additional structured details about this error.
  //
  // Keys should match /[a-zA-Z0-9-_]/ and be limited to 64 characters in
  // length. When identifying the current value of an exceeded limit, the units
  // should be contained in the key, not the value.  For example, rather than
  // {"instanceLimit": "100/request"}, should be returned as,
  // {"instanceLimitPerRequest": "100"}, if the client exceeds the number of
  // instances that can be created in a single (batch) request.
  map<string, string> metadata = 3;
}
```

### Error Status
The properties of the **google.rpc.Status** will be populated as following:

> - **Code**  - Protocol level error code. These could be either gRPC or HTTP error codes. See (gRPC Codes ProtoBuf)[https://github.com/googleapis/googleapis/blob/master/google/rpc/code.proto]

Author: We could address the HTTP error codes with a mapping like the one shown in the table: https://cloud.google.com/apis/design/errors#handling_errors

> - **Code**  - Protocol level error code. These could be either gRPC or HTTP error codes. See (gRPC Codes ProtoBuf)[https://github.com/googleapis/googleapis/blob/master/google/rpc/code.proto]

Reviewer 3: How are errors returned in HTTP, in general?

> - **Code**  - Protocol level error code. These could be either gRPC or HTTP error codes. See (gRPC Codes ProtoBuf)[https://github.com/googleapis/googleapis/blob/master/google/rpc/code.proto]

Author: @ItalyPaleAle  Here what I did as part of my POC, but of course it could implemented better: https://github.com/robertojrojas/dapr/blob/a0760ea549b9a3da93af2bee96e88ad94db9b32b/pkg/http/api.go#L1017

   
   Example: "InvalidArgument Code = 3", "Internal Code = 13"

- **Message**  - Error message.
- **Details**  - A set of standard error payloads for error details. These list can be found in [Error Details](https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto)
   
   Example: "ErrorInfo", "ResourceInfo"

Below is partial table of the Standard Error code provided by gRPC and how they map to HTTP error codes. The entire list can found in the following links:
- [Google API Error Handling]https://cloud.google.com/apis/design/errors#handling_errors
- (gRPC Codes ProtoBuf)[https://github.com/googleapis/googleapis/blob/master/google/rpc/code.proto]

|HTTP | gRPC      | Description |
|---- | ----------- | ----------- |
|200  |	OK		  | No error. |
|400  |	INVALID_ARGUMENT |	Client specified an invalid argument. Check error message and error details for more information. |
|400  |	FAILED_PRECONDITION |	Request can not be executed in the current system state, such as deleting a non-empty directory. |
|400  |	OUT_OF_RANGE |	Client specified an invalid range. |

> |401  |	UNAUTHENTICATED	| Request not authenticated due to missing, invalid, or expired OAuth token. |
> |403  |	PERMISSION_DENIED |	Client does not have sufficient permission. This can happen because the OAuth token does not have the right scopes, the client doesn't have permission, or the API has not been enabled. |

Reviewer 3: Nit: we don't use OAuth.

```suggestion
|401  |	UNAUTHENTICATED	| Request not authenticated due to missing, invalid, or expired authorization credentials. |
|403  |	PERMISSION_DENIED |	Client does not have sufficient permission. |
```

|404  |	NOT_FOUND |	A specified resource is not found. |
|409  |	ABORTED	| Concurrency conflict, such as read-modify-write conflict. |

### ErrorInfo (Required)
The properties of the **type.googleapis.com/google.rpc.ErrorInfo** will be populated as following:

- **Reason** - A combination of prefix from prefix of the table below plus the error condition code.
    
    Example: "DAPR_STATE_" + "ETAG_MISMATCH"

- **Domain** - With the value `dapr.io`.

- **Metadata** - A key/value map/dictionary data relevant to the error condition.

`**Note:**` The metadata property **isRetriable** with a truthable value("true", "false", "True", "False", "TRUE", "FALSE", "1", "0") is required. 

### ResourceInfo (Optional)
The properties of the **type.googleapis.com/google.rpc.ResourceInfo** will be populated as following:

- **ResourceType** - The building block type.
   
   Example: "StateStore"

- **ResourceName** - The component name.
   

>    Example: "REDIS"

Author: I like `state.redis/v1` for `ResourceName`. 
@artursouza any thoughts?

>    Example: "REDIS"

Reviewer 3: Great callout.

Either a different key or `state.redis/v1/component-name`?

>    Example: "REDIS"

Reviewer 2: I am fine with `state.redis/v1/component-name`

- **Owner**   - The owner of the component.

- **Description** - Resource descrpition or error details.

### Error Details Prefixes
The following tables shows the propsosed error codes prefixes used in the **reason** for the **google.rpc.ErrorInfo** for various Dapr building blocks:

**INIT**
| Dapr Module | Prefix |
| ----------- | ----------- |
| CLI         | DAPR_CLI_INIT_* |
| Self-hosted | DAPR_SELF_HOSTED_INIT_* |
| K8S         | DAPR_K8S_INIT_* |
| Invoke      | DAPR_INVOKE_INIT_* |

**RUNTIME**
| Dapr Module | Prefix |
| ----------- | ----------- |
| CLI         | DAPR_CLI_RUNTIME_* |
| Self-hosted | DAPR_SELF_HOSTED_RUNTIME_* |

> | dapr-2-dapr(HTTP) | DAPR_HTTP_RUNTIME_* |

Author: @ItalyPaleAle This is one of those part of Darp I'm not totally clear. I added the HTTP protocol for Dapr-2-Dapr thinking that Dapr side cars have an option of replacing gRPC for HTTP for internal communication of each other. isn't that the case?

> | dapr-2-dapr(HTTP) | DAPR_HTTP_RUNTIME_* |

Reviewer 3: Internal communication (sidecar-to-sidecar) always happens over gRPC.

| dapr-2-dapr(gRPC) | DAPR_GRPC_RUNTIME_* |

**COMPONENTS**
| Dapr Module | Prefix |
| ----------- | ----------- |
| PubSub              | DAPR_PUBSUB_* |
| StateStore          | DAPR_STATE_* |
| Bindings            | DAPR_BINDING_* |
| SecretStore         | DAPR_SECRET_* |
| ConfigurationStore  | DAPR_CONFIGURATION_* |
| Lock                | DAPR_LOCK_* |
| NameResolution      | DAPR_NAME_RESOLUTION_* |
| Middleware          | DAPR_MIDDLEWARE_*|

The following snippet shows an error status returned due to a `ETAG_MISMATCH` error condition. The **reason** is populated with `PREFIX+ERROR_CONDITION`:

```json
{
  "code": 3,
  "message":  "possible etag mismatch. error from state store",
  "details": [
    {
      "@type": "type.googleapis.com/google.rpc.ErrorInfo",
      "reason": "DAPR_STATE_ETAG_MISMATCH",
      "domain": "dapr.io",
      "metadata": {
        "key": "myapp||name"
      }
    },
    {
      "@type": "type.googleapis.com/google.rpc.ResourceInfo",

>       "resource_type": "StateStore",

Reviewer 3: Some keys use camelCase, some use snake_case. We should be consistent

Also see my comment above. I think a single `resource_name` = `state.redis/v1` may be more aligned with how we refer to the comoponents everywhere (note that some components have a dot in the name too, such as `state.azure.cosmosdb`, and we should include the version too)

Author: The `keys`  for the JSON payload are auto-generated as part of the Marshaling implemented by the Protobuf stub of the [error_details.proto](https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto) 

if you look at the annotations, you'll see the `resource_type` as the JSON key to generate/read.
```
ResourceType string `protobuf:"bytes,1,opt,name=resource_type,json=resourceType,proto3" json:"resource_type,omitempty"`
```

|HTTP | gRPC      | Description |
|---- | ----------- | ----------- |
|200  |	OK		  | No error. |
|400  |	INVALID_ARGUMENT |	Client specified an invalid argument. Check error message and error details for more information. |
|400  |	FAILED_PRECONDITION |	Request can not be executed in the current system state, such as deleting a non-empty directory. |
|400  |	OUT_OF_RANGE |	Client specified an invalid range. |
|401  |	UNAUTHENTICATED	| Request not authenticated due to missing, invalid, or expired OAuth token. |
|403  |	PERMISSION_DENIED |	Client does not have sufficient permission. This can happen because the OAuth token does not have the right scopes, the client doesn't have permission, or the API has not been enabled. |
|404  |	NOT_FOUND |	A specified resource is not found. |
|409  |	ABORTED	| Concurrency conflict, such as read-modify-write conflict. |

### Error Details Classification
The following tables shows the propsosed error codes used in the **reason** for the **google.rpc.ErrorInfo** within the various levels of Dapr modules:

**INIT**
| Dapr Module | Description |
| ----------- | ----------- |

> | CLI         | DAPR_CLI_INIT |

Reviewer 2: Prefix only: DAPR_CLI_*

| Self-hosted | DAPR_SELF_HOSTED_INIT |
| K8S         | DAPR_K8S_INIT |
| Invoke      | DAPR_INVOKE_INIT |

**RUNTIME**
| Dapr Module | Description |
| ----------- | ----------- |
| CLI         | DAPR_CLI_RUNTIME |
| Self-hosted | DAPR_SELF_HOSTED_RUNTIME |
| dapr-2-dapr(HTTP) | DAPR_HTTP_RUNTIME |
| dapr-2-dapr(gRPC) | DAPR_GRPC_RUNTIME |

**COMPONENTS**
| Dapr Module | Description |
| ----------- | ----------- |
| PubSub              | DAPR_PUBSUB_COMPONENT |

> | StateStore          | DAPR_STATE_STORE_COMPONENT |

Reviewer 2: Same here, DAPR_STATE_*

| Bindings            | DAPR_BINDING_COMPONENT |
| SecretStore         | DAPR_SECRET_STORE_COMPONENT |
| ConfigurationStore  | DAPR_CONFIGURATION_STORE_COMPONENT |
| Lock                | DAPR_LOCK_COMPONENT |
| NameResolution      | DAPR_NAME_RESOLUTION_COMPONENT |
| Middleware          | DAPR_MIDDLEWARE_COMPONENT|

### Sample Error
```json
{
  "code": 8089,
  "message": "unable to perform Redis.HGETALL call: dial tcp [::1]:6379: connect: connection refused",
  "details": [
    {
      "@type": "type.googleapis.com/google.rpc.ErrorInfo",

>       "reason": "DAPR_STATE_STORE_COMPONENT",

Reviewer 2: The reason should be more specific for the client to make use of it. For example, was it a eTag mismatch or a random failure trying to write?

Reviewer 2: DAPR_STATE_ETAG_MISMATCH

|HTTP | gRPC      | Description |
|---- | ----------- | ----------- |
|200  |	OK		  | No error. |
|400  |	INVALID_ARGUMENT |	Client specified an invalid argument. Check error message and error details for more information. |
|400  |	FAILED_PRECONDITION |	Request can not be executed in the current system state, such as deleting a non-empty directory. |
|400  |	OUT_OF_RANGE |	Client specified an invalid range. |
|401  |	UNAUTHENTICATED	| Request not authenticated due to missing, invalid, or expired OAuth token. |
|403  |	PERMISSION_DENIED |	Client does not have sufficient permission. This can happen because the OAuth token does not have the right scopes, the client doesn't have permission, or the API has not been enabled. |
|404  |	NOT_FOUND |	A specified resource is not found. |
|409  |	ABORTED	| Concurrency conflict, such as read-modify-write conflict. |

### Error Details Classification
The following tables shows the propsosed error codes used in the **reason** for the **google.rpc.ErrorInfo** within the various levels of Dapr modules:

**INIT**
| Dapr Module | Description |
| ----------- | ----------- |

> | CLI         | DPR_INIT_CLI_[1-1000] |

Reviewer 2: nit: just spell out DAPR since it is not saving much.

> | Self-hosted | DPR_INIT_SLF_[1-1000] |

Reviewer 2: Unless we really need the space, can we just spell it out? If we need to save space, we can truncate the message.

| K8S | DPR_INIT_K8S_[1-1000] |
| Invoke | DPR_INIT_INV_[1-1000] |

**RUNTIME**
| Dapr Module | Description |
| ----------- | ----------- |
| CLI         | DPR_RT_CLI_[1-1000] |
| Self-hosted | DPR_RT_SLF_[1-1000] |
| dapr-2-dapr(HTTP) | DPR_INIT_HTTP_[1-1000] |
| dapr-2-dapr(gRPC) | DPR_INIT_GRPC_[1-1000] |

**COMPONENTS**
| Dapr Module | Description |
| ----------- | ----------- |
| PubSub              | DPR_COMP_SUB_[1-1000] |
| StateStore          | DPR_COMP_STS_[1-1000] |
| Bindings            | DPR_COMP_BND_[1-1000] |
| SecretStore         | DPR_COMP_SEC_[1-1000] |
| ConfigurationStore  | DPR_COMP_CNF_[1-1000] |
| Lock                | DPR_COMP_LCK_[1-1000] |
| NameResolution      | DPR_COMP_NMR_[1-1000] |
| Middleware          | DPR_COMP_MDW_[1-1000] |

### Sample Error
```json
{
  "code": 8089,
  "message": "unable to perform Redis.HGETALL call: dial tcp [::1]:6379: connect: connection refused",
  "details": [
    {
      "@type": "type.googleapis.com/google.rpc.ErrorInfo",
      "reason": "DPR_COMP_STS_1",

>       "domain": "redis.state.contrib.dapr.io",

Reviewer 2: I think the domain should not contain the component type. It feels that there is enough information in the reason that would make the domain simply be "dapr.io" - see how reason has "STS" and domain has "state", reason has "COMP" and domain has "contrib".

      "metadata": {
        "key": "myapp||name"
      }
    },
    {
      "@type": "type.googleapis.com/google.rpc.ResourceInfo",
      "resource_type": "StateStore",

>       "resource_name": "REDIS",

Reviewer 2: This is good, it means we can keep ErrorInfo agnostic of component implementation.

  "details": [
    {
      "@type": "type.googleapis.com/google.rpc.ErrorInfo",
      "reason": "DPR_COMP_STS_1",
      "domain": "redis.state.contrib.dapr.io",
      "metadata": {
        "key": "myapp||name"
      }
    },
    {
      "@type": "type.googleapis.com/google.rpc.ResourceInfo",
      "resource_type": "StateStore",
      "resource_name": "REDIS",
      "owner": "",
      "description": "unable to perform Redis.HGETALL call: dial tcp [::1]:6379: connect: connection refused"
    },
    {
      "@type": "type.googleapis.com/kit.proto.customerrors.v1.DaprKitErrorInfo",
      "reason": "DPR_COMP_STS_1",
      "domain": "redis.state.contrib.dapr.io",
      "metadata": {
        "key": "myapp||name"
      }
    }
  ]
}
```
### Pros
- Since the Dapr Runtime is using protocol buffers as the data format, support for the richer error model is already included in most of the gRPC implementations.
- This would help minimize the changes with the Dapr ecosystem.
- This solution could be used to programmatically react to errors as it provides a standard structure for the errors with details.

### Cons
- Dependencies on gPRC richer error model.

> - Not all gRPC implementations support this out of the box, [C#/dotnet](https://learn.microsoft.com/en-us/dotnet/architecture/grpc-for-wcf-developers/error-handling#grpc-richer-error-model) being one of them, but there is a [possible solution (Google.Api.CommonProtos)](https://www.nuget.org/packages/Google.Api.CommonProtos/) to this.

Reviewer 2: This is a big deal, but we need to confirm with @halspang if the extra dependency is OK.

Author: `dotnet-sdk` works well. I need to test `js-sdk`, `java-sdk`, and `python-sdk`

## gRPC Richer Error Model POC
For the POC I've made changes to some parts of the Dapr modules (). The POC code can be found in my GH Repo under the branch **error-codes-poc**

These are the gRPC imports used:

```go
import (
  ...
  "google.golang.org/genproto/googleapis/rpc/errdetails"
  "google.golang.org/grpc/codes"
  "google.golang.org/grpc/status"
  ...
)
```

The files changed for this POC:

https://github.com/robertojrojas/components-contrib/tree/error-codes-poc

- state/redis/redis.go
- state/store.go

https://github.com/robertojrojas/dapr-kit/tree/error-codes-poc

- pkg/proto/customerrors/v1/customerrors.pb.go
- proto/customerrors/v1/customerrors.proto
- status/customerrorcodes.go
- status/status.go

https://github.com/robertojrojas/dapr/tree/error-codes-poc

- pkg/diagnostics/grpc_tracing.go
- pkg/grpc/api.go
- pkg/http/api.go
- pkg/http/responses.go

https://github.com/robertojrojas/dapr-go-sdk/tree/error-codes-poc

- client/state.go

https://github.com/robertojrojas/dapr-cli/tree/error-codes-poc

- pkg/standalone/invoke.go

> ## Solution B

Reviewer 2: Mention this as alternative consideration but the proposal should have only one solution and not two.
